tealer/analyses/utils/stack_emulator.py

```diff
@@ -0,0 +1,451 @@
+
+    Instructions to access information of recently submitted inner transaction:
+        itxn f, itxna f i, gitxn t f, gitxna t f i, itxnas f, gitxnas t f,
+
+
+Stack Manipulation Opcodes:
+
+- Instructions which change the order of elements in the stack independent of the actual values of the elements.
+- could be emulated during stack emulation
+
+    bury n, popn n, dupn n, pop, dup,
+    dup2, dig n, swap, cover n, uncover n,
+
+    select, # select depends on the runtime information
+
+
+Scratch Space:
+    Instructions to manipulate scratch space:
+        load, store, loads, stores,
+
+    Instructions to access scratch space of previous transactions in the group:
+        gload t i, gloads i, gloadss
+
+
+Control Flow:
+    (Might) Stop the execution:
+        err, return, assert,
+
+    Branch, Jump around the code:
+        b target
+        bnz target, bz target, switch, match,
+
+        callsub, retsub,
+
+
+Arthimetic Operations:
+    +, -, /, *, %, sqrt, exp, shl, shr,
+    addw, mulw, divw, divmodw, expw,        # wide math
+    b+, b-, b/, b*, b%, bsqrt,              # byte (512-bits)
+
+
+Bitwise Operations:
+    |, &, ^, ~,
+    b|, b&, b^, b~,         # byte (512-bits)
+
+
+Connectors:
+    &&
+    ||
+
+Comparison
+    <, >, <=, >=, ==, !=, !
+    b<, b>, b<=, b>=, b==, b!=
+
+
+Int/Byte Manipulation:
+    bzero,
+    concat,
+    len, bitlen,
+    itob, btoi,
+    getbit, setbit, getbyte, setbyte,
+    substring s e, substring3,
+    extract, extract3, extract_uint16, extract_uint32, extract_uint64,
+    replace2, replace3,
+    base64_decode, json_ref,
+
+
+Cryptographic/Hash:
+    sha256, keccak256, sha512_256, sha3_256,
+    ed25519verify, ed25519verify_bare,
+    ecdsa_verify, ecdsa_pk_decompress, ecdsa_pk_recover
+    vrf_verify,
+
+
+Logging Operations:
+    log
+
+
+Unclear/recent(v8) instructions:
+    proto, frame_dig, frame_bury,
+"""
+
+from typing import List, Union, Dict, Type, Tuple
+from functools import lru_cache
+
+
+from tealer.teal.basic_blocks import BasicBlock
+from tealer.teal.instructions.instructions import (
+    Instruction,
+)
+
+
+class UnknownStackValue:
+    """Represent a unknown stack value."""
+
+    def __str__(self) -> str:
+        return "UnknownStackValue()"
+
+    def __repr__(self) -> str:
+        return "UnknownStackValue()"
+
+
+class KnownStackValue:
+    """Represent values which are known during the local stack emulation.
+
+    A stack value is represented using the instruction, arguments and the position.
+
+    instruction: Every stack value is a result of executing a instruction. instruction
+        property contains the reference to that instruction's object.
+    args: A instruction consumes values from the stack. `args` represents the list of stack values
+        which are consumed by `self.instruction` to produce the result(s).
+    position: A instruction might produce multiple values. position represents the index in the list
+        of produced values.
```

Comment:
Maybe we should mention explicitly that this is only the result of intermediate push. "might produce multiple values" is a bit confusing

Reply:
I would explicit state that this is only for push, and we are not building intermediate stack for pop